README.md
# ESP32 PWM Controller

PWM Controller using ESP32 and PCA9685, cabable of controlling multiple Servos via I2C.

## Hardware

* [ESP32-D1-mini](https://www.az-delivery.de/products/esp32-d1-mini)
* [PCA9685](https://cdn-shop.adafruit.com/datasheets/PCA9685.pdf)

## Dependencies

* [ESP-IDF](https://github.com/espressif/esp-idf) >= v4.4

## Build

See installation guide for Windows
* [docs.espressif/windows-setup](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/get-started/windows-setup.html)
* [docs.espressif/windows-setup-update](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/get-started/windows-setup-update.html)

```powershell
# Open Powershell
# ESP_IDF_PATH --> path to esp-idf directory
$ESP_IDF_PATH\export.ps1
idf.py build
```
## Run main application

```powershell
# Open Powershell
# WORKSPACE --> path to project directory
$ESP_IDF_PATH\export.ps1
cd $WORKSPACE
idf.py -p $PORT flash
idf.py -p $PORT monitor
```

## Run test cases

```powershell
# Open Powershell
$ESP_IDF_PATH\export.ps1
cd $WORKSPACE\test
idf.py -p $PORT flash
idf.py -p $PORT monitor
```

CHANGELOG.md
## [0.2.0] - 2022-04-02

### Added

- basic functionality working

   - write8, read8
   - wakeup, sleep, restart
   - set_duty_cycle, set_refresh_freq

-----------------------------------------------------------

## [0.1.0] - 2022-01-09

### Added

- Dummy main application

- Added component pca9685

   - PCA9685 class
   - Register mapping
   - Read/Write register functions + tests
   - Set PWM & Frequency functions + tests

test/main/main.c
/**
 * @brief       main.c
 * @details     Main test application
 * @return      NA
 *
 * @author      
 * @date        
 * @version     
 */
#include <stdio.h>
#include <string.h>
#include "unity.h"
#include "freertos/FreeRTOS.h"
#include "freertos/task.h"
#include "esp_system.h"

void app_main(void)
{
  UNITY_BEGIN();
  unity_run_all_tests();
  UNITY_END();
}


components/pca9685/test/test_rw.cpp
/**
 * @brief       test_rw.c
 * @details     Test read/write
 * @return      NA
 *
 * @author      
 * @date        
 * @version     
 */
#include <limits.h>
#include "unity.h"
#include "esp_log.h"
#include "freertos/FreeRTOS.h"
#include "freertos/task.h"
#include "driver/i2c.h"
#include "pca9685.h"

const gpio_num_t i2c_sda = GPIO_NUM_21;
const gpio_num_t i2c_scl = GPIO_NUM_22;
#define I2C_ADDR 0x70
#define I2C_REG_ADDR 0x06

TEST_CASE("Write/Read register with main api", "[pca9685]")
{
  // initialization
  esp_err_t ret;
  i2c_config_t conf;
  conf.mode = I2C_MODE_MASTER;
  conf.sda_io_num = i2c_sda;
  conf.scl_io_num = i2c_scl;
  conf.sda_pullup_en = GPIO_PULLUP_ENABLE;
  conf.scl_pullup_en = GPIO_PULLUP_ENABLE;
  conf.clk_flags = I2C_SCLK_SRC_FLAG_FOR_NOMAL;
  conf.master.clk_speed = 400000;
  ESP_ERROR_CHECK(i2c_param_config(I2C_NUM_0, &conf));
  ESP_ERROR_CHECK(i2c_driver_install(I2C_NUM_0, conf.mode, 0, 0, 0));
  // Wait for initialization to finish
  vTaskDelay(100 / portTICK_PERIOD_MS);

  // read/write 
  uint8_t data = 0x17;
  uint8_t write_buf[2] = {I2C_REG_ADDR, data};
  uint8_t read_buf = 0x00;
  ret = i2c_master_write_to_device(I2C_NUM_0, I2C_ADDR, write_buf, sizeof(write_buf), 1000 / portTICK_RATE_MS);
  TEST_ASSERT(ret == ESP_OK);
  ret = i2c_master_read_from_device(I2C_NUM_0, I2C_ADDR, &read_buf, sizeof(read_buf), 1000 / portTICK_RATE_MS);
  TEST_ASSERT(ret == ESP_OK);
  TEST_ASSERT(data == read_buf);
  ret = i2c_driver_delete(I2C_NUM_0);
  TEST_ASSERT(ret == ESP_OK);
}

TEST_CASE("Write/Read register with class wrapper - single byte", "[pca9685]")
{
  uint8_t data_w = 0x17;
  PCA9685 *pca9685 = new PCA9685(I2C_ADDR, i2c_sda, i2c_scl);

  pca9685->write8(I2C_REG_ADDR, data_w);
  uint8_t data_r = pca9685->read8(I2C_REG_ADDR);

  TEST_ASSERT(data_w == data_r);
  pca9685->~PCA9685();
}


components/pca9685/test/test_pwm_freq.cpp
/**
 * @brief       test_pwm_freq.cpp
 * @details     Test setting of frequency and duty cycle
 * @return      NA
 *
 * @author      
 * @date        
 * @version     
 */
#include <limits.h>
#include "unity.h"
#include "esp_log.h"
#include "freertos/FreeRTOS.h"
#include "freertos/task.h"
#include "driver/i2c.h"
#include "pca9685.h"


const gpio_num_t i2c_sda = GPIO_NUM_21;
const gpio_num_t i2c_scl = GPIO_NUM_22;
#define I2C_ADDR 0x70


TEST_CASE("Set duty cycle, LED0", "[pca9685]")
{
  PCA9685 *pca9685 = new PCA9685(I2C_ADDR, i2c_sda, i2c_scl);
  pca9685->set_duty_cycle(0, 0, PCA9685::MAX_PWM_COUNT/2);
  pca9685->~PCA9685();
}

TEST_CASE("Set refresh frequency", "[pca9685]")
{
  PCA9685 *pca9685 = new PCA9685(I2C_ADDR, i2c_sda, i2c_scl);
  pca9685->set_refresh_freq(500);
  pca9685->~PCA9685();
}


components/pca9685/pca9685.h
/**
 * @brief       PCA9685.h
 * @details     16-channel, 12-bit PWM I2C-bus controller.
 * @return      NA
 *
 * @author
 * @date
 * @version
 */

#pragma once

#include "driver/i2c.h"


class PCA9685
{
  public:
    // Registers addresses
    static const uint8_t MODE1          = 0x00;  /*  Mode register 1 */
    static const uint8_t MODE2          = 0x01;  /*  Mode register 2 */
    static const uint8_t SUBADR1        = 0x02;  /*  I2C-bus subaddress 1 */
    static const uint8_t SUBADR2        = 0x03;  /*  I2C-bus subaddress 2 */
    static const uint8_t SUBADR3        = 0x04;  /*  I2C-bus subaddress 3 */
    static const uint8_t ALLCALLADR     = 0x05;  /*  LED All Call I2C-bus address */
    static const uint8_t LED0_ON_L      = 0x06;  /*  LED0 output and brightness control byte 0 */
    static const uint8_t LED0_ON_H      = 0x07;  /*  LED0 output and brightness control byte 1 */
    static const uint8_t LED0_OFF_L     = 0x08;  /*  LED0 output and brightness control byte 2 */
    static const uint8_t LED0_OFF_H     = 0x09;  /*  LED0 output and brightness control byte 3 */
    static const uint8_t ALL_LED_ON_L   = 0xFA;  /*  ALL_LED output and brightness control byte 0 */
    static const uint8_t ALL_LED_ON_H   = 0xFB;  /*  ALL_LED output and brightness control byte 1 */
    static const uint8_t ALL_LED_OFF_L  = 0xFC;  /*  ALL_LED output and brightness control byte 2 */
    static const uint8_t ALL_LED_OFF_H  = 0xFD;  /*  ALL_LED output and brightness control byte 3 */
    static const uint8_t PRE_SCALE      = 0xFE;  /*  prescaler for PWM output frequency */
    static const uint8_t TESTMODE       = 0xFF;  /*  defines the test mode to be entered */
    // MODE1 bits
    static const uint8_t MODE1_ALLCAL   = 0;  /* respond to LED All Call I2C-bus address */
    static const uint8_t MODE1_SUB3     = 1;  /* respond to I2C-bus subaddress 3 */
    static const uint8_t MODE1_SUB2     = 2;  /* respond to I2C-bus subaddress 2 */
    static const uint8_t MODE1_SUB1     = 3;  /* respond to I2C-bus subaddress 1 */
    static const uint8_t MODE1_SLEEP    = 4;  /* Low power mode. Oscillator off */
    static const uint8_t MODE1_AI       = 5;  /* Auto-Increment enabled */
    static const uint8_t MODE1_EXTCLK   = 6;  /* Use EXTCLK pin clock */
    static const uint8_t MODE1_RESTART  = 7;  /* Restart enabled */
    // MODE2 bits
    static const uint8_t MODE2_OUTNE_0  = 0;  /* Active LOW output enable input */
    static const uint8_t MODE2_OUTNE_1  = 1;  /* Active LOW output enable input - high impedience */
    static const uint8_t MODE2_OUTDRV   = 2;  /* totem pole structure vs open-drain */
    static const uint8_t MODE2_OCH      = 3;  /* Outputs change on ACK vs STOP */
    static const uint8_t MODE2_INVRT    = 4;  /* Output logic state inverted */
    // Defaults & Constants
    static const uint16_t MAX_REFRESH_RATE  = 1526;     /* Maximum refresh rate, Hz */
    static const uint16_t MIN_REFRESH_RATE  = 24;       /* Minimum refresh rate, Hz */
    static const uint16_t MAX_PWM_COUNT     = 4095;     /* Maximum value of 12 bit counter */
    static const uint8_t LED_STEP_FACTOR    = 4;        /* Factor to calculate next LED */
    static const uint32_t CLK_FREQ          = 25000000; /* PCA9685 internal clock frequency */
    static const uint32_t REFRESH_FREQ      = 200;      /* Default PWM frequency */
    // I2C Settings
    static const uint8_t I2C_NUM          = I2C_NUM_0; /* Default I2C bus */
    static const uint8_t I2C_ADDR         = 0x70;      /* Default PCA9685 I2C Slave Address */
    static const uint32_t I2C_CLK_SPEED   = 400000;    /* Default master clock speed */
    static const uint32_t I2C_TIMEOUT_MS  = 1000;      /* Default timeout_ms */
    // I2C User Settings
    uint32_t i2c_num;         /* User-specified I2C bus */
    uint32_t i2c_addr;        /* User-specified PCA9685 I2C Slave Address */
    uint32_t i2c_clk_speed;   /* User-specified master clock speed */
    uint32_t i2c_timeout_ms;  /* User-specified timeout_ms */

    PCA9685(uint32_t i2c_num, uint32_t i2c_addr, gpio_num_t i2c_sda, gpio_num_t i2c_scl, uint32_t i2c_clk_speed, uint32_t i2c_timeout_ms);
    PCA9685(uint32_t i2c_addr, gpio_num_t i2c_sda, gpio_num_t i2c_scl);
    ~PCA9685();
    void write8(uint8_t i2c_reg_addr, uint8_t data);
    uint8_t read8(uint8_t i2c_reg_addr);
    void set_duty_cycle(uint8_t led, uint16_t on, uint16_t off);
    void set_refresh_freq(uint16_t frequency);
    void wakeup();
    void sleep();
    void restart();

  private:
    uint8_t toggle_bit(uint8_t num, uint8_t bit_nr);
    uint8_t set_bit(uint8_t num, uint8_t bit_nr);
    uint8_t clear_bit(uint8_t num, uint8_t bit_nr);
    uint8_t check_bit(uint8_t num, uint8_t bit_nr);
};


components/pca9685/pca9685.cpp
/**
 * @brief       PCA9685.c
 * @details     16-channel, 12-bit PWM I2C-bus controller.
 * @return      NA
 *
 * @author      
 * @date        
 * @version     
 */
#include <stdio.h>
#include <math.h>
#include "esp_log.h"
#include "freertos/FreeRTOS.h"
#include "freertos/task.h"
#include "driver/i2c.h"
#include "pca9685.h"


//************************************************************************************************ PRIVATE

uint8_t PCA9685::toggle_bit(uint8_t num, uint8_t bit_nr)   {return num ^=  (1 << bit_nr);}
uint8_t PCA9685::set_bit(uint8_t num, uint8_t bit_nr)      {return num |=  (1 << bit_nr);}
uint8_t PCA9685::clear_bit(uint8_t num, uint8_t bit_nr)    {return num ^=  (1 << bit_nr);}
uint8_t PCA9685::check_bit(uint8_t num, uint8_t bit_nr)    {return num &   (1 << bit_nr);}


//************************************************************************************************ PUBLIC

PCA9685::PCA9685(uint32_t i2c_num, uint32_t i2c_addr, gpio_num_t i2c_sda, gpio_num_t i2c_scl, uint32_t i2c_clk_speed, uint32_t i2c_timeout_ms)
{
  PCA9685::i2c_num = i2c_num;
  PCA9685::i2c_addr = i2c_addr;
  PCA9685::i2c_clk_speed = i2c_clk_speed;
  PCA9685::i2c_timeout_ms = i2c_timeout_ms;

  i2c_config_t conf;
  conf.mode = I2C_MODE_MASTER;
  conf.sda_io_num = i2c_sda;
  conf.scl_io_num = i2c_scl;
  conf.sda_pullup_en = GPIO_PULLUP_ENABLE;
  conf.scl_pullup_en = GPIO_PULLUP_ENABLE;
  conf.clk_flags = I2C_SCLK_SRC_FLAG_FOR_NOMAL;
  conf.master.clk_speed = i2c_clk_speed;
  ESP_ERROR_CHECK_WITHOUT_ABORT(i2c_param_config(i2c_num, &conf));
  ESP_ERROR_CHECK_WITHOUT_ABORT(i2c_driver_install(i2c_num, conf.mode, 0, 0, 0));
  vTaskDelay(100 / portTICK_PERIOD_MS);
}

PCA9685::PCA9685(uint32_t i2c_addr, gpio_num_t i2c_sda, gpio_num_t i2c_scl)
{
  PCA9685::i2c_num = I2C_NUM;
  PCA9685::i2c_addr = i2c_addr;
  PCA9685::i2c_clk_speed = I2C_CLK_SPEED;
  PCA9685::i2c_timeout_ms = I2C_TIMEOUT_MS;

  i2c_config_t conf;
  conf.mode = I2C_MODE_MASTER;
  conf.sda_io_num = i2c_sda;
  conf.scl_io_num = i2c_scl;
  conf.sda_pullup_en = GPIO_PULLUP_ENABLE;
  conf.scl_pullup_en = GPIO_PULLUP_ENABLE;
  conf.clk_flags = I2C_SCLK_SRC_FLAG_FOR_NOMAL;
  conf.master.clk_speed = I2C_CLK_SPEED;
  ESP_ERROR_CHECK_WITHOUT_ABORT(i2c_param_config(I2C_NUM, &conf));
  ESP_ERROR_CHECK_WITHOUT_ABORT(i2c_driver_install(I2C_NUM, conf.mode, 0, 0, 0));
  vTaskDelay(100 / portTICK_PERIOD_MS);
}

PCA9685::~PCA9685()
{
  ESP_ERROR_CHECK_WITHOUT_ABORT(i2c_driver_delete(i2c_num));
}

void PCA9685::write8(uint8_t i2c_reg_addr, uint8_t data)
{
  uint8_t buffer[2] = {i2c_reg_addr, data};
  ESP_ERROR_CHECK_WITHOUT_ABORT(i2c_master_write_to_device(i2c_num, i2c_addr, buffer, 2, i2c_timeout_ms / portTICK_RATE_MS));
}

uint8_t PCA9685::read8(uint8_t i2c_reg_addr)
{
  uint8_t buffer = i2c_reg_addr;
  uint8_t data = 0x00;
  ESP_ERROR_CHECK_WITHOUT_ABORT(i2c_master_write_to_device(i2c_num, i2c_addr, &buffer, 1, i2c_timeout_ms / portTICK_RATE_MS));
  ESP_ERROR_CHECK_WITHOUT_ABORT(i2c_master_read_from_device(i2c_num, i2c_addr, &data, 1, i2c_timeout_ms / portTICK_RATE_MS));
  return data;
}

void PCA9685::set_duty_cycle(uint8_t led, uint16_t on, uint16_t off)
{
  sleep();

  // limits
  if (on  > MAX_PWM_COUNT) on = MAX_PWM_COUNT;
  if (off > MAX_PWM_COUNT) off = MAX_PWM_COUNT;

  // split uint16_t values to H and L bytes
  uint8_t ON_L  = (uint8_t) (on & 0x00FF);
  uint8_t ON_H  = (uint8_t) ((on & 0xFF00) >> 8);
  uint8_t OFF_L = (uint8_t) (off & 0x00FF);
  uint8_t OFF_H = (uint8_t) ((off & 0xFF00) >> 8);

  // calculate offset for LEDx, x > 0
  uint8_t offset = LED_STEP_FACTOR * (uint8_t)led;

  write8(LED0_ON_L + offset, ON_L);
  write8(LED0_ON_H + offset, ON_H);
  write8(LED0_OFF_L + offset, OFF_L);
  write8(LED0_OFF_H + offset, OFF_H);

  // wakeup and restart
  uint8_t mode1_reg = read8(MODE1);
  if(check_bit(mode1_reg, MODE1_RESTART))
  {
    wakeup();
  }
  restart();
}

void PCA9685::set_refresh_freq(uint16_t frequency)
{
  sleep();

  // limits
  if(frequency < MIN_REFRESH_RATE) {
    frequency = MIN_REFRESH_RATE;
  } else if(frequency > MAX_REFRESH_RATE) {
    frequency = MAX_REFRESH_RATE;
  }

  // calculate prescale value
  double pres = round(double(CLK_FREQ) / ((double)frequency * 4096)) - 1;
  uint8_t prescale = (uint8_t)pres;

  write8(PRE_SCALE, prescale);

  // wakeup
  wakeup();
}

void PCA9685::wakeup()
{
  uint8_t state = read8(MODE1);
  uint8_t wake_mode = clear_bit(state, MODE1_SLEEP); 
  write8(MODE1, wake_mode);
  vTaskDelay(10 / portTICK_PERIOD_MS);
}

void PCA9685::sleep()
{
  uint8_t state = read8(MODE1);
  uint8_t sleep_mode = set_bit(state, MODE1_SLEEP); 
  write8(MODE1, sleep_mode);
  vTaskDelay(10 / portTICK_PERIOD_MS);
}

void PCA9685::restart()
{
  uint8_t state = read8(MODE1);
  uint8_t restart = set_bit(state, MODE1_RESTART);
  write8(MODE1, restart);
  vTaskDelay(10 / portTICK_PERIOD_MS);
}
